// @vicinae/api/dist/commands/lint/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const node_fs_1 = require("node:fs");
const node_path_1 = require("node:path");
const logger_js_1 = require("../../utils/logger.js");
const manifest_js_1 = __importDefault(require("../../schemas/manifest.js"));
class Lint extends core_1.Command {
    static args = {};
    static description = "Validate the extension manifest (package.json)";
    static examples = [
        `<%= config.bin %> <%= command.id %>`,
        `<%= config.bin %> <%= command.id %> --src /path/to/extension`,
    ];
    static flags = {
        src: core_1.Flags.string({
            aliases: ["src"],
            char: "s",
            default: process.cwd(),
            defaultHelp: "The current working directory",
            description: "Path to the extension source directory",
            required: false,
        }),
    };
    async run() {
        const { flags } = await this.parse(Lint);
        const logger = new logger_js_1.Logger();
        const src = flags.src ?? process.cwd();
        const pkgPath = (0, node_path_1.join)(src, "package.json");
        if (!(0, node_fs_1.existsSync)(pkgPath)) {
            logger.logError(`No package.json found at ${pkgPath}. Does this location point to a valid extension repository?`);
            process.exit(1);
        }
        const json = JSON.parse((0, node_fs_1.readFileSync)(pkgPath, "utf8"));
        const result = manifest_js_1.default.safeParse(json);
        if (result.error) {
            logger.logError(`${pkgPath} is not a valid extension manifest: ${result.error}`);
            process.exit(1);
        }
        logger.logReady(`Manifest is valid`);
    }
}
exports.default = Lint;
