// @vicinae/api/dist/api/proto/oauth.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.7.7
//   protoc               v6.33.1
// source: oauth.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Response = exports.Request = exports.AuthorizeResponse = exports.RemoveTokensResponse = exports.RemoveTokensRequest = exports.SetTokensResponse = exports.SetTokensRequest = exports.GetTokensResponse = exports.GetTokensRequest = exports.TokenSet = exports.AuthorizeRequest = exports.PKCEClientOptions = exports.RedirectMethod = exports.protobufPackage = void 0;
exports.redirectMethodFromJSON = redirectMethodFromJSON;
exports.redirectMethodToJSON = redirectMethodToJSON;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const ui_1 = require("./ui");
exports.protobufPackage = "proto.ext.oauth";
var RedirectMethod;
(function (RedirectMethod) {
    RedirectMethod[RedirectMethod["Web"] = 0] = "Web";
    RedirectMethod[RedirectMethod["App"] = 1] = "App";
    RedirectMethod[RedirectMethod["AppURI"] = 2] = "AppURI";
    RedirectMethod[RedirectMethod["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(RedirectMethod || (exports.RedirectMethod = RedirectMethod = {}));
function redirectMethodFromJSON(object) {
    switch (object) {
        case 0:
        case "Web":
            return RedirectMethod.Web;
        case 1:
        case "App":
            return RedirectMethod.App;
        case 2:
        case "AppURI":
            return RedirectMethod.AppURI;
        case -1:
        case "UNRECOGNIZED":
        default:
            return RedirectMethod.UNRECOGNIZED;
    }
}
function redirectMethodToJSON(object) {
    switch (object) {
        case RedirectMethod.Web:
            return "Web";
        case RedirectMethod.App:
            return "App";
        case RedirectMethod.AppURI:
            return "AppURI";
        case RedirectMethod.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBasePKCEClientOptions() {
    return { id: undefined, name: "", description: "", icon: undefined };
}
exports.PKCEClientOptions = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.id !== undefined) {
            writer.uint32(10).string(message.id);
        }
        if (message.name !== "") {
            writer.uint32(18).string(message.name);
        }
        if (message.description !== "") {
            writer.uint32(26).string(message.description);
        }
        if (message.icon !== undefined) {
            ui_1.Image.encode(message.icon, writer.uint32(34).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePKCEClientOptions();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.description = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.icon = ui_1.Image.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? globalThis.String(object.id) : undefined,
            name: isSet(object.name) ? globalThis.String(object.name) : "",
            description: isSet(object.description) ? globalThis.String(object.description) : "",
            icon: isSet(object.icon) ? ui_1.Image.fromJSON(object.icon) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== undefined) {
            obj.id = message.id;
        }
        if (message.name !== "") {
            obj.name = message.name;
        }
        if (message.description !== "") {
            obj.description = message.description;
        }
        if (message.icon !== undefined) {
            obj.icon = ui_1.Image.toJSON(message.icon);
        }
        return obj;
    },
    create(base) {
        return exports.PKCEClientOptions.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBasePKCEClientOptions();
        message.id = object.id ?? undefined;
        message.name = object.name ?? "";
        message.description = object.description ?? "";
        message.icon = (object.icon !== undefined && object.icon !== null) ? ui_1.Image.fromPartial(object.icon) : undefined;
        return message;
    },
};
function createBaseAuthorizeRequest() {
    return { client: undefined, url: "" };
}
exports.AuthorizeRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.client !== undefined) {
            exports.PKCEClientOptions.encode(message.client, writer.uint32(10).fork()).join();
        }
        if (message.url !== "") {
            writer.uint32(18).string(message.url);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAuthorizeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.client = exports.PKCEClientOptions.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.url = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            client: isSet(object.client) ? exports.PKCEClientOptions.fromJSON(object.client) : undefined,
            url: isSet(object.url) ? globalThis.String(object.url) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.client !== undefined) {
            obj.client = exports.PKCEClientOptions.toJSON(message.client);
        }
        if (message.url !== "") {
            obj.url = message.url;
        }
        return obj;
    },
    create(base) {
        return exports.AuthorizeRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAuthorizeRequest();
        message.client = (object.client !== undefined && object.client !== null)
            ? exports.PKCEClientOptions.fromPartial(object.client)
            : undefined;
        message.url = object.url ?? "";
        return message;
    },
};
function createBaseTokenSet() {
    return {
        accessToken: "",
        refreshToken: undefined,
        idToken: undefined,
        expiresIn: undefined,
        scope: undefined,
        updatedAt: 0,
    };
}
exports.TokenSet = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.accessToken !== "") {
            writer.uint32(10).string(message.accessToken);
        }
        if (message.refreshToken !== undefined) {
            writer.uint32(18).string(message.refreshToken);
        }
        if (message.idToken !== undefined) {
            writer.uint32(26).string(message.idToken);
        }
        if (message.expiresIn !== undefined) {
            writer.uint32(32).uint32(message.expiresIn);
        }
        if (message.scope !== undefined) {
            writer.uint32(42).string(message.scope);
        }
        if (message.updatedAt !== 0) {
            writer.uint32(48).uint64(message.updatedAt);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTokenSet();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.accessToken = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.refreshToken = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.idToken = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.expiresIn = reader.uint32();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.scope = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 48) {
                        break;
                    }
                    message.updatedAt = longToNumber(reader.uint64());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            accessToken: isSet(object.accessToken) ? globalThis.String(object.accessToken) : "",
            refreshToken: isSet(object.refreshToken) ? globalThis.String(object.refreshToken) : undefined,
            idToken: isSet(object.idToken) ? globalThis.String(object.idToken) : undefined,
            expiresIn: isSet(object.expiresIn) ? globalThis.Number(object.expiresIn) : undefined,
            scope: isSet(object.scope) ? globalThis.String(object.scope) : undefined,
            updatedAt: isSet(object.updatedAt) ? globalThis.Number(object.updatedAt) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.accessToken !== "") {
            obj.accessToken = message.accessToken;
        }
        if (message.refreshToken !== undefined) {
            obj.refreshToken = message.refreshToken;
        }
        if (message.idToken !== undefined) {
            obj.idToken = message.idToken;
        }
        if (message.expiresIn !== undefined) {
            obj.expiresIn = Math.round(message.expiresIn);
        }
        if (message.scope !== undefined) {
            obj.scope = message.scope;
        }
        if (message.updatedAt !== 0) {
            obj.updatedAt = Math.round(message.updatedAt);
        }
        return obj;
    },
    create(base) {
        return exports.TokenSet.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseTokenSet();
        message.accessToken = object.accessToken ?? "";
        message.refreshToken = object.refreshToken ?? undefined;
        message.idToken = object.idToken ?? undefined;
        message.expiresIn = object.expiresIn ?? undefined;
        message.scope = object.scope ?? undefined;
        message.updatedAt = object.updatedAt ?? 0;
        return message;
    },
};
function createBaseGetTokensRequest() {
    return { providerId: undefined };
}
exports.GetTokensRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.providerId !== undefined) {
            writer.uint32(10).string(message.providerId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetTokensRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.providerId = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { providerId: isSet(object.providerId) ? globalThis.String(object.providerId) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.providerId !== undefined) {
            obj.providerId = message.providerId;
        }
        return obj;
    },
    create(base) {
        return exports.GetTokensRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetTokensRequest();
        message.providerId = object.providerId ?? undefined;
        return message;
    },
};
function createBaseGetTokensResponse() {
    return { tokenSet: undefined };
}
exports.GetTokensResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.tokenSet !== undefined) {
            exports.TokenSet.encode(message.tokenSet, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetTokensResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.tokenSet = exports.TokenSet.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { tokenSet: isSet(object.tokenSet) ? exports.TokenSet.fromJSON(object.tokenSet) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.tokenSet !== undefined) {
            obj.tokenSet = exports.TokenSet.toJSON(message.tokenSet);
        }
        return obj;
    },
    create(base) {
        return exports.GetTokensResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetTokensResponse();
        message.tokenSet = (object.tokenSet !== undefined && object.tokenSet !== null)
            ? exports.TokenSet.fromPartial(object.tokenSet)
            : undefined;
        return message;
    },
};
function createBaseSetTokensRequest() {
    return {
        providerId: undefined,
        accessToken: "",
        refreshToken: undefined,
        idToken: undefined,
        expiresIn: undefined,
        scope: undefined,
    };
}
exports.SetTokensRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.providerId !== undefined) {
            writer.uint32(10).string(message.providerId);
        }
        if (message.accessToken !== "") {
            writer.uint32(18).string(message.accessToken);
        }
        if (message.refreshToken !== undefined) {
            writer.uint32(26).string(message.refreshToken);
        }
        if (message.idToken !== undefined) {
            writer.uint32(34).string(message.idToken);
        }
        if (message.expiresIn !== undefined) {
            writer.uint32(40).uint32(message.expiresIn);
        }
        if (message.scope !== undefined) {
            writer.uint32(50).string(message.scope);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSetTokensRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.providerId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.accessToken = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.refreshToken = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.idToken = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 40) {
                        break;
                    }
                    message.expiresIn = reader.uint32();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.scope = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            providerId: isSet(object.providerId) ? globalThis.String(object.providerId) : undefined,
            accessToken: isSet(object.accessToken) ? globalThis.String(object.accessToken) : "",
            refreshToken: isSet(object.refreshToken) ? globalThis.String(object.refreshToken) : undefined,
            idToken: isSet(object.idToken) ? globalThis.String(object.idToken) : undefined,
            expiresIn: isSet(object.expiresIn) ? globalThis.Number(object.expiresIn) : undefined,
            scope: isSet(object.scope) ? globalThis.String(object.scope) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.providerId !== undefined) {
            obj.providerId = message.providerId;
        }
        if (message.accessToken !== "") {
            obj.accessToken = message.accessToken;
        }
        if (message.refreshToken !== undefined) {
            obj.refreshToken = message.refreshToken;
        }
        if (message.idToken !== undefined) {
            obj.idToken = message.idToken;
        }
        if (message.expiresIn !== undefined) {
            obj.expiresIn = Math.round(message.expiresIn);
        }
        if (message.scope !== undefined) {
            obj.scope = message.scope;
        }
        return obj;
    },
    create(base) {
        return exports.SetTokensRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseSetTokensRequest();
        message.providerId = object.providerId ?? undefined;
        message.accessToken = object.accessToken ?? "";
        message.refreshToken = object.refreshToken ?? undefined;
        message.idToken = object.idToken ?? undefined;
        message.expiresIn = object.expiresIn ?? undefined;
        message.scope = object.scope ?? undefined;
        return message;
    },
};
function createBaseSetTokensResponse() {
    return {};
}
exports.SetTokensResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSetTokensResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.SetTokensResponse.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseSetTokensResponse();
        return message;
    },
};
function createBaseRemoveTokensRequest() {
    return { providerId: undefined };
}
exports.RemoveTokensRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.providerId !== undefined) {
            writer.uint32(10).string(message.providerId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoveTokensRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.providerId = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { providerId: isSet(object.providerId) ? globalThis.String(object.providerId) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.providerId !== undefined) {
            obj.providerId = message.providerId;
        }
        return obj;
    },
    create(base) {
        return exports.RemoveTokensRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRemoveTokensRequest();
        message.providerId = object.providerId ?? undefined;
        return message;
    },
};
function createBaseRemoveTokensResponse() {
    return {};
}
exports.RemoveTokensResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoveTokensResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.RemoveTokensResponse.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseRemoveTokensResponse();
        return message;
    },
};
function createBaseAuthorizeResponse() {
    return { code: "" };
}
exports.AuthorizeResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.code !== "") {
            writer.uint32(10).string(message.code);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAuthorizeResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.code = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { code: isSet(object.code) ? globalThis.String(object.code) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.code !== "") {
            obj.code = message.code;
        }
        return obj;
    },
    create(base) {
        return exports.AuthorizeResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAuthorizeResponse();
        message.code = object.code ?? "";
        return message;
    },
};
function createBaseRequest() {
    return { authorize: undefined, getTokens: undefined, setTokens: undefined, removeTokens: undefined };
}
exports.Request = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.authorize !== undefined) {
            exports.AuthorizeRequest.encode(message.authorize, writer.uint32(10).fork()).join();
        }
        if (message.getTokens !== undefined) {
            exports.GetTokensRequest.encode(message.getTokens, writer.uint32(18).fork()).join();
        }
        if (message.setTokens !== undefined) {
            exports.SetTokensRequest.encode(message.setTokens, writer.uint32(26).fork()).join();
        }
        if (message.removeTokens !== undefined) {
            exports.RemoveTokensRequest.encode(message.removeTokens, writer.uint32(34).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.authorize = exports.AuthorizeRequest.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.getTokens = exports.GetTokensRequest.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.setTokens = exports.SetTokensRequest.decode(reader, reader.uint32());
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.removeTokens = exports.RemoveTokensRequest.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            authorize: isSet(object.authorize) ? exports.AuthorizeRequest.fromJSON(object.authorize) : undefined,
            getTokens: isSet(object.getTokens) ? exports.GetTokensRequest.fromJSON(object.getTokens) : undefined,
            setTokens: isSet(object.setTokens) ? exports.SetTokensRequest.fromJSON(object.setTokens) : undefined,
            removeTokens: isSet(object.removeTokens) ? exports.RemoveTokensRequest.fromJSON(object.removeTokens) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.authorize !== undefined) {
            obj.authorize = exports.AuthorizeRequest.toJSON(message.authorize);
        }
        if (message.getTokens !== undefined) {
            obj.getTokens = exports.GetTokensRequest.toJSON(message.getTokens);
        }
        if (message.setTokens !== undefined) {
            obj.setTokens = exports.SetTokensRequest.toJSON(message.setTokens);
        }
        if (message.removeTokens !== undefined) {
            obj.removeTokens = exports.RemoveTokensRequest.toJSON(message.removeTokens);
        }
        return obj;
    },
    create(base) {
        return exports.Request.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRequest();
        message.authorize = (object.authorize !== undefined && object.authorize !== null)
            ? exports.AuthorizeRequest.fromPartial(object.authorize)
            : undefined;
        message.getTokens = (object.getTokens !== undefined && object.getTokens !== null)
            ? exports.GetTokensRequest.fromPartial(object.getTokens)
            : undefined;
        message.setTokens = (object.setTokens !== undefined && object.setTokens !== null)
            ? exports.SetTokensRequest.fromPartial(object.setTokens)
            : undefined;
        message.removeTokens = (object.removeTokens !== undefined && object.removeTokens !== null)
            ? exports.RemoveTokensRequest.fromPartial(object.removeTokens)
            : undefined;
        return message;
    },
};
function createBaseResponse() {
    return { authorize: undefined, getTokens: undefined, setTokens: undefined, removeTokens: undefined };
}
exports.Response = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.authorize !== undefined) {
            exports.AuthorizeResponse.encode(message.authorize, writer.uint32(10).fork()).join();
        }
        if (message.getTokens !== undefined) {
            exports.GetTokensResponse.encode(message.getTokens, writer.uint32(18).fork()).join();
        }
        if (message.setTokens !== undefined) {
            exports.SetTokensResponse.encode(message.setTokens, writer.uint32(26).fork()).join();
        }
        if (message.removeTokens !== undefined) {
            exports.RemoveTokensResponse.encode(message.removeTokens, writer.uint32(34).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.authorize = exports.AuthorizeResponse.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.getTokens = exports.GetTokensResponse.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.setTokens = exports.SetTokensResponse.decode(reader, reader.uint32());
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.removeTokens = exports.RemoveTokensResponse.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            authorize: isSet(object.authorize) ? exports.AuthorizeResponse.fromJSON(object.authorize) : undefined,
            getTokens: isSet(object.getTokens) ? exports.GetTokensResponse.fromJSON(object.getTokens) : undefined,
            setTokens: isSet(object.setTokens) ? exports.SetTokensResponse.fromJSON(object.setTokens) : undefined,
            removeTokens: isSet(object.removeTokens) ? exports.RemoveTokensResponse.fromJSON(object.removeTokens) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.authorize !== undefined) {
            obj.authorize = exports.AuthorizeResponse.toJSON(message.authorize);
        }
        if (message.getTokens !== undefined) {
            obj.getTokens = exports.GetTokensResponse.toJSON(message.getTokens);
        }
        if (message.setTokens !== undefined) {
            obj.setTokens = exports.SetTokensResponse.toJSON(message.setTokens);
        }
        if (message.removeTokens !== undefined) {
            obj.removeTokens = exports.RemoveTokensResponse.toJSON(message.removeTokens);
        }
        return obj;
    },
    create(base) {
        return exports.Response.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseResponse();
        message.authorize = (object.authorize !== undefined && object.authorize !== null)
            ? exports.AuthorizeResponse.fromPartial(object.authorize)
            : undefined;
        message.getTokens = (object.getTokens !== undefined && object.getTokens !== null)
            ? exports.GetTokensResponse.fromPartial(object.getTokens)
            : undefined;
        message.setTokens = (object.setTokens !== undefined && object.setTokens !== null)
            ? exports.SetTokensResponse.fromPartial(object.setTokens)
            : undefined;
        message.removeTokens = (object.removeTokens !== undefined && object.removeTokens !== null)
            ? exports.RemoveTokensResponse.fromPartial(object.removeTokens)
            : undefined;
        return message;
    },
};
function longToNumber(int64) {
    const num = globalThis.Number(int64.toString());
    if (num > globalThis.Number.MAX_SAFE_INTEGER) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    if (num < globalThis.Number.MIN_SAFE_INTEGER) {
        throw new globalThis.Error("Value is smaller than Number.MIN_SAFE_INTEGER");
    }
    return num;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
