// @vicinae/api/dist/utils/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const yoctocolors_1 = require("./yoctocolors");
class Logger {
    prefixes = {
        error: `${(0, yoctocolors_1.red)("error")}${(0, yoctocolors_1.reset)()}`,
        event: `${(0, yoctocolors_1.magenta)("event")}${(0, yoctocolors_1.reset)()}`,
        info: `${(0, yoctocolors_1.blue)("info")}${(0, yoctocolors_1.reset)()}`,
        ready: `${(0, yoctocolors_1.green)("ready")}${(0, yoctocolors_1.reset)()}`,
    };
    logError(message) {
        console.log(`${this.prefixes.error.padEnd(15)} - ${message}`);
    }
    logEvent(message) {
        console.log(`${this.prefixes.event.padEnd(15)} - ${message}`);
    }
    logInfo(message) {
        console.log(`${this.prefixes.info.padEnd(15)} - ${message}`);
    }
    logReady(message) {
        console.log(`${this.prefixes.ready.padEnd(15)} - ${message}`);
    }
    logExtensionOut(s) {
        this.logTimestamp(s);
    }
    logExtensionError(s) {
        this.logTimestamp(`${(0, yoctocolors_1.red)(s)}${(0, yoctocolors_1.reset)()}`);
    }
    logTimestamp(s) {
        const ts = new Date().toJSON();
        const lines = s.split("\n");
        for (let i = 0; i !== lines.length; ++i) {
            const line = lines[i];
            if (i === lines.length - 1 && line.length === 0)
                continue;
            console.log(`${(0, yoctocolors_1.gray)(ts.padEnd(20))}${(0, yoctocolors_1.reset)()} - ${line}`);
        }
    }
}
exports.Logger = Logger;
