// @vicinae/api/dist/utils/yoctocolors.js
"use strict";
// https://github.com/sindresorhus/yoctocolors
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bgWhiteBright = exports.bgCyanBright = exports.bgMagentaBright = exports.bgBlueBright = exports.bgYellowBright = exports.bgGreenBright = exports.bgRedBright = exports.whiteBright = exports.cyanBright = exports.magentaBright = exports.blueBright = exports.yellowBright = exports.greenBright = exports.redBright = exports.bgGray = exports.bgWhite = exports.bgCyan = exports.bgMagenta = exports.bgBlue = exports.bgYellow = exports.bgGreen = exports.bgRed = exports.bgBlack = exports.gray = exports.white = exports.cyan = exports.magenta = exports.blue = exports.yellow = exports.green = exports.red = exports.black = exports.strikethrough = exports.hidden = exports.inverse = exports.overline = exports.underline = exports.italic = exports.dim = exports.bold = exports.reset = void 0;
const node_tty_1 = __importDefault(require("node:tty"));
// eslint-disable-next-line no-warning-comments
// TODO: Use a better method when it's added to Node.js (https://github.com/nodejs/node/pull/40240)
// Lots of optionals here to support Deno.
const hasColors = node_tty_1.default?.WriteStream?.prototype?.hasColors?.() ?? false;
const format = (open, close) => {
    if (!hasColors) {
        return (input) => input ?? "";
    }
    const openCode = `\u001B[${open}m`;
    const closeCode = `\u001B[${close}m`;
    return (input = "") => {
        let index = input.indexOf(closeCode);
        if (index === -1) {
            // Note: Intentionally not using string interpolation for performance reasons.
            return openCode + input + closeCode;
        }
        // Handle nested colors.
        // We could have done this, but it's too slow (as of Node.js 22).
        // return openCode + string.replaceAll(closeCode, (close === 22 ? closeCode : '') + openCode) + closeCode;
        let result = openCode;
        let lastIndex = 0;
        // SGR 22 resets both bold (1) and dim (2). When we encounter a nested
        // close for styles that use 22, we need to re-open the outer style.
        const reopenOnNestedClose = close === 22;
        const replaceCode = (reopenOnNestedClose ? closeCode : "") + openCode;
        while (index !== -1) {
            result += input.slice(lastIndex, index) + replaceCode;
            lastIndex = index + closeCode.length;
            index = input.indexOf(closeCode, lastIndex);
        }
        result += input.slice(lastIndex) + closeCode;
        return result;
    };
};
exports.reset = format(0, 0);
exports.bold = format(1, 22);
exports.dim = format(2, 22);
exports.italic = format(3, 23);
exports.underline = format(4, 24);
exports.overline = format(53, 55);
exports.inverse = format(7, 27);
exports.hidden = format(8, 28);
exports.strikethrough = format(9, 29);
exports.black = format(30, 39);
exports.red = format(31, 39);
exports.green = format(32, 39);
exports.yellow = format(33, 39);
exports.blue = format(34, 39);
exports.magenta = format(35, 39);
exports.cyan = format(36, 39);
exports.white = format(37, 39);
exports.gray = format(90, 39);
exports.bgBlack = format(40, 49);
exports.bgRed = format(41, 49);
exports.bgGreen = format(42, 49);
exports.bgYellow = format(43, 49);
exports.bgBlue = format(44, 49);
exports.bgMagenta = format(45, 49);
exports.bgCyan = format(46, 49);
exports.bgWhite = format(47, 49);
exports.bgGray = format(100, 49);
exports.redBright = format(91, 39);
exports.greenBright = format(92, 39);
exports.yellowBright = format(93, 39);
exports.blueBright = format(94, 39);
exports.magentaBright = format(95, 39);
exports.cyanBright = format(96, 39);
exports.whiteBright = format(97, 39);
exports.bgRedBright = format(101, 49);
exports.bgGreenBright = format(102, 49);
exports.bgYellowBright = format(103, 49);
exports.bgBlueBright = format(104, 49);
exports.bgMagentaBright = format(105, 49);
exports.bgCyanBright = format(106, 49);
exports.bgWhiteBright = format(107, 49);
