// @vicinae/api/dist/api/proto/command.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.7.7
//   protoc               v6.32.1
// source: command.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Response = exports.Request = exports.UpdateCommandMetadataRequest = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const common_1 = require("./common");
exports.protobufPackage = "proto.ext.command";
function createBaseUpdateCommandMetadataRequest() {
    return { subtitle: undefined };
}
exports.UpdateCommandMetadataRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.subtitle !== undefined) {
            writer.uint32(10).string(message.subtitle);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUpdateCommandMetadataRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.subtitle = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { subtitle: isSet(object.subtitle) ? globalThis.String(object.subtitle) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.subtitle !== undefined) {
            obj.subtitle = message.subtitle;
        }
        return obj;
    },
    create(base) {
        return exports.UpdateCommandMetadataRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseUpdateCommandMetadataRequest();
        message.subtitle = object.subtitle ?? undefined;
        return message;
    },
};
function createBaseRequest() {
    return { updateCommandMetadata: undefined };
}
exports.Request = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.updateCommandMetadata !== undefined) {
            exports.UpdateCommandMetadataRequest.encode(message.updateCommandMetadata, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.updateCommandMetadata = exports.UpdateCommandMetadataRequest.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            updateCommandMetadata: isSet(object.updateCommandMetadata)
                ? exports.UpdateCommandMetadataRequest.fromJSON(object.updateCommandMetadata)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.updateCommandMetadata !== undefined) {
            obj.updateCommandMetadata = exports.UpdateCommandMetadataRequest.toJSON(message.updateCommandMetadata);
        }
        return obj;
    },
    create(base) {
        return exports.Request.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRequest();
        message.updateCommandMetadata =
            (object.updateCommandMetadata !== undefined && object.updateCommandMetadata !== null)
                ? exports.UpdateCommandMetadataRequest.fromPartial(object.updateCommandMetadata)
                : undefined;
        return message;
    },
};
function createBaseResponse() {
    return { updateCommandMetadata: undefined };
}
exports.Response = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.updateCommandMetadata !== undefined) {
            common_1.AckResponse.encode(message.updateCommandMetadata, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.updateCommandMetadata = common_1.AckResponse.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            updateCommandMetadata: isSet(object.updateCommandMetadata)
                ? common_1.AckResponse.fromJSON(object.updateCommandMetadata)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.updateCommandMetadata !== undefined) {
            obj.updateCommandMetadata = common_1.AckResponse.toJSON(message.updateCommandMetadata);
        }
        return obj;
    },
    create(base) {
        return exports.Response.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseResponse();
        message.updateCommandMetadata =
            (object.updateCommandMetadata !== undefined && object.updateCommandMetadata !== null)
                ? common_1.AckResponse.fromPartial(object.updateCommandMetadata)
                : undefined;
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
