// @vicinae/api/dist/api/components/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Action = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const node_crypto_1 = require("node:crypto");
const index_1 = require("../hooks/index");
const clipboard_1 = require("../clipboard");
const image_1 = require("../image");
const utils_1 = require("../utils");
const icon_1 = require("../icon");
const controls_1 = require("../controls");
const ActionRoot = ({ icon, ...props }) => {
    const serializedIcon = icon ? (0, image_1.serializeProtoImage)(icon) : icon;
    const stableIdRef = (0, react_1.useRef)(undefined);
    if (!stableIdRef.current) {
        stableIdRef.current = (0, node_crypto_1.randomUUID)();
    }
    return ((0, jsx_runtime_1.jsx)("action", { ...props, icon: serializedIcon, stableId: stableIdRef.current }));
};
const CopyToClipboard = ({ title = "Copy to clipboard", icon = icon_1.Icon.CopyClipboard, content, concealed = false, onCopy, ...props }) => {
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { title: title, ...props, icon: icon_1.Icon.CopyClipboard, onAction: async () => {
            clipboard_1.Clipboard.copy(content, { concealed });
            (0, controls_1.closeMainWindow)();
            onCopy?.(content);
        } }));
};
const Paste = ({ title = "Paste to active window", icon = icon_1.Icon.CopyClipboard, content, onPaste, ...props }) => {
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { title: title, ...props, icon: icon_1.Icon.CopyClipboard, onAction: async () => {
            (0, controls_1.closeMainWindow)(); // we close before pasting to make sure focus has been properly restored
            clipboard_1.Clipboard.paste(content);
            onPaste?.(content);
        } }));
};
const Open = ({ target, app, ...props }) => {
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { ...props, onAction: () => {
            (0, utils_1.open)(target, app);
        } }));
};
const OpenInBrowser = ({ url, title = "Open in Browser", ...props }) => {
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { ...props, title: title, onAction: async () => {
            await (0, utils_1.open)(url);
            await (0, controls_1.closeMainWindow)();
        } }));
};
const ShowInFinder = ({ path, title = "Show in Finder", ...props }) => {
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { ...props, title: title, onAction: () => {
            (0, utils_1.showInFileBrowser)(path);
        } }));
};
const Push = ({ target, ...props }) => {
    const { push } = (0, index_1.useNavigation)();
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { ...props, onAction: () => {
            push(target);
        } }));
};
const SubmitForm = ({ title = "Submit", ...props }) => {
    const stableIdRef = (0, react_1.useRef)(undefined);
    if (!stableIdRef.current) {
        stableIdRef.current = (0, node_crypto_1.randomUUID)();
    }
    const nativeProps = {
        ...props,
        stableId: stableIdRef.current,
        title,
        icon: props.icon ? (0, image_1.serializeProtoImage)(props.icon) : props.icon,
        onAction: () => { },
    };
    return (0, jsx_runtime_1.jsx)("action", { ...nativeProps });
};
const CreateQuicklink = ({ title = "Create Quicklink", quicklink, ...props }) => {
    const stableIdRef = (0, react_1.useRef)(undefined);
    if (!stableIdRef.current) {
        stableIdRef.current = (0, node_crypto_1.randomUUID)();
    }
    const nativeProps = {
        ...props,
        stableId: stableIdRef.current,
        title,
        type: "create-quicklink",
        quicklink: {
            link: quicklink.link,
            name: quicklink.name,
            application: typeof quicklink.application === "string"
                ? quicklink.application
                : quicklink.application?.name,
            icon: quicklink.icon,
        },
        icon: props.icon ? (0, image_1.serializeProtoImage)(props.icon) : props.icon,
        onAction: () => { },
    };
    return (0, jsx_runtime_1.jsx)("action", { ...nativeProps });
};
/**
 * @category Actions
 */
exports.Action = Object.assign(ActionRoot, {
    CopyToClipboard,
    Push,
    Open,
    Paste,
    SubmitForm,
    OpenInBrowser,
    ShowInFinder,
    CreateQuicklink,
    Style: {
        Regular: "regular",
        Destructive: "destructive",
    },
});
