// @vicinae/api/dist/api/components/grid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Grid = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const image_1 = require("../image");
const crypto_1 = require("crypto");
const empty_view_1 = require("./empty-view");
const color_1 = require("../color");
const dropdown_1 = require("./dropdown");
var GridInset;
(function (GridInset) {
    GridInset["Zero"] = "zero";
    GridInset["Small"] = "small";
    GridInset["Medium"] = "medium";
    GridInset["Large"] = "large";
})(GridInset || (GridInset = {}));
/**
 * Enum representing the number of items that should be displayed on a single row.
 * @deprecated - use `columns` instead.
 */
var GridItemSize;
(function (GridItemSize) {
    GridItemSize["Small"] = "small";
    GridItemSize["Medium"] = "medium";
    GridItemSize["Large"] = "large";
})(GridItemSize || (GridItemSize = {}));
const aspectRatioMap = {
    "1": 1,
    "3/2": 3 / 2,
    "2/3": 2 / 3,
    "4/3": 4 / 3,
    "3/4": 3 / 4,
    "16/9": 16 / 9,
    "9/16": 9 / 16,
    "21/9": 21 / 9,
    "9/21": 9 / 21,
    "32/9": 32 / 9,
    "9/32": 9 / 32,
};
/**
 * Controls how the grid content should fit its allocated space.
 */
var GridFit;
(function (GridFit) {
    /**
     * The content will be contained within the grid cell, with vertical/horizontal bars if its aspect ratio differs from the cell's
     */
    GridFit["Contain"] = "contain";
    /**
     * The content will be scaled proportionally so that it fill the entire cell; parts of the content could end up being cropped out.
     */
    GridFit["Fill"] = "fill";
})(GridFit || (GridFit = {}));
const GridRoot = ({ searchBarAccessory, children, actions, inset, itemSize, fit = GridFit.Contain, aspectRatio = "1", ...props }) => {
    if (typeof props.enableFiltering === "boolean" &&
        typeof props.filtering === "undefined") {
        props.filtering = props.enableFiltering;
    }
    if (props.columns === undefined && itemSize) {
        props.columns = {
            [GridItemSize.Small]: 8,
            [GridItemSize.Medium]: 5,
            [GridItemSize.Large]: 3,
        }[itemSize];
    }
    return ((0, jsx_runtime_1.jsxs)("grid", { fit: fit, inset: inset, aspectRatio: aspectRatioMap[aspectRatio], ...props, children: [searchBarAccessory, children, actions] }));
};
const GridItem = ({ detail, actions, content, accessory, ...props }) => {
    const id = (0, react_1.useRef)(props.id ?? (0, crypto_1.randomUUID)());
    // Remove value wrapper
    const innerContent = typeof content === "object" && "value" in content ? content.value : content;
    const tooltip = typeof content === "object" && "tooltip" in content
        ? content.tooltip
        : undefined;
    // Content
    let serializedContent;
    if (innerContent &&
        typeof innerContent === "object" &&
        "color" in innerContent) {
        serializedContent = {
            color: (0, color_1.serializeColorLike)(innerContent.color),
        };
    }
    else {
        serializedContent = innerContent
            ? (0, image_1.serializeProtoImage)(innerContent)
            : undefined;
    }
    // Accessory
    const serializedAccessory = accessory
        ? {
            icon: accessory.icon
                ? (0, image_1.serializeProtoImage)(accessory.icon)
                : accessory.icon,
            tooltip: accessory.tooltip,
        }
        : undefined;
    return ((0, jsx_runtime_1.jsxs)("grid-item", { ...props, content: serializedContent, tooltip: tooltip, accessory: serializedAccessory, id: id.current, children: [detail, actions] }));
};
const GridSection = ({ fit, aspectRatio, inset, ...props }) => {
    const nativeProps = {
        fit,
        aspectRatio: aspectRatio ? aspectRatioMap[aspectRatio] : aspectRatio,
        inset,
        ...props,
    };
    return (0, jsx_runtime_1.jsx)("grid-section", { ...nativeProps });
};
/**
 * @category UI Components
 */
exports.Grid = Object.assign(GridRoot, {
    Section: GridSection,
    EmptyView: empty_view_1.EmptyView,
    Dropdown: dropdown_1.Dropdown,
    Fit: GridFit,
    Inset: GridInset,
    ItemSize: GridItemSize,
    Item: Object.assign(GridItem, {}),
});
