// @vicinae/api/dist/commands/develop/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const chokidar = __importStar(require("chokidar"));
const esbuild = __importStar(require("esbuild"));
const node_child_process_1 = require("node:child_process");
const fs = __importStar(require("node:fs"));
const fsp = __importStar(require("node:fs/promises"));
const path = __importStar(require("node:path"));
const logger_js_1 = require("../../utils/logger.js");
const utils_js_1 = require("../../utils/utils.js");
const extension_types_js_1 = require("../../utils/extension-types.js");
const vicinae_js_1 = require("../../utils/vicinae.js");
const manifest_js_1 = __importDefault(require("../../schemas/manifest.js"));
const tail_js_1 = require("../../utils/tail.js");
class Develop extends core_1.Command {
    static args = {};
    static description = "Start an extension development session";
    static examples = [
        `<%= config.bin %> <%= command.id %> --target /path/to/extension`,
    ];
    static flags = {
        target: core_1.Flags.string({
            aliases: ["input"],
            char: "i",
            default: process.cwd(),
            defaultHelp: "The current working directory",
            description: "Path to the extension directory",
            required: false,
        }),
    };
    async run() {
        const { flags } = await this.parse(Develop);
        const logger = new logger_js_1.Logger();
        const pkgPath = path.join(flags.target, "package.json");
        const parseManifest = () => {
            if (!fs.existsSync(pkgPath)) {
                logger.logError(`No package.json found at ${pkgPath}. Does this location point to a valid extension repository?`);
                process.exit(1);
            }
            const json = JSON.parse(fs.readFileSync(pkgPath, "utf8"));
            const e = manifest_js_1.default.safeParse(json);
            if (e.error) {
                logger.logError(`${pkgPath} is not a valid extension manifest: ${e.error}`);
                process.exit(1);
            }
            return e.data;
        };
        let manifest = parseManifest();
        const vicinae = new vicinae_js_1.VicinaeClient();
        logger.logInfo("Generating extension types...");
        (0, extension_types_js_1.updateExtensionTypes)(manifest, flags.target);
        const typeCheck = async () => {
            const spawned = (0, node_child_process_1.spawn)("npx", ["tsc", "--noEmit"]);
            let stderr = Buffer.from("");
            return new Promise((resolve) => {
                spawned.stderr.on("data", (buf) => {
                    stderr = Buffer.concat([stderr, buf]);
                });
                spawned.on("exit", (status) => resolve({ error: stderr.toString(), ok: status === 0 }));
            });
        };
        const build = async (outDir) => {
            /*
      logger.logInfo("Started type checking in background thread");
      typeCheck().then(({ error, ok }) => {
        if (!ok) {
          logger.logInfo(`Type checking error: ${error}`);
        }

        logger.logInfo("Done type checking");
      });
      */
            const entryPoints = manifest.commands
                .map((cmd) => path.join("src", `${cmd.name}.tsx`))
                .filter(fs.existsSync);
            logger.logInfo(`entrypoints [${entryPoints.join(", ")}]`);
            const promises = manifest.commands.map((cmd) => {
                const base = path.join(process.cwd(), "src", `${cmd.name}`);
                const tsxSource = `${base}.tsx`;
                const tsSource = `${base}.ts`;
                let source = tsxSource;
                if (cmd.mode === "view" && !fs.existsSync(tsxSource)) {
                    throw new Error(`could not find entrypoint src/${cmd.name}.tsx for command ${cmd.name}.`);
                }
                // we allow .ts or .tsx for no-view
                if (cmd.mode === "no-view") {
                    if (!fs.existsSync(tsxSource)) {
                        source = tsSource;
                        if (!fs.existsSync(tsSource)) {
                            throw new Error(`could not find entrypoint src/${cmd.name}.{ts,tsx} for command ${cmd.name}.`);
                        }
                    }
                }
                return esbuild.build({
                    bundle: true,
                    entryPoints: [source],
                    external: ["react", "@vicinae/api", "@raycast/api"],
                    format: "cjs",
                    outfile: path.join(outDir, `${cmd.name}.js`),
                    platform: "node",
                });
            });
            await Promise.all(promises);
            const targetPkg = path.join(outDir, "package.json");
            const targetAssets = path.join(outDir, "assets");
            fs.cpSync("package.json", targetPkg, { force: true });
            if (fs.existsSync("assets")) {
                fs.cpSync("assets", targetAssets, { force: true, recursive: true });
            }
            else {
                fs.mkdirSync(targetAssets, { recursive: true });
            }
        };
        const pingError = vicinae.ping();
        if (pingError) {
            console.error(`Failed to ping vicinae\n`, pingError.message);
            return;
        }
        const safeBuild = async (extensionDir) => {
            try {
                const start = performance.now();
                await build(extensionDir);
                const time = performance.now() - start;
                logger.logReady(`Extension built in ${Math.round(time)}ms 🚀`);
                vicinae.refreshDevSession(id);
            }
            catch (error) {
                if (error instanceof Error) {
                    logger.logError(`Failed to build extension: ${error.message}`);
                }
                else {
                    logger.logError(`Failed to build extension: ${error}`);
                }
            }
        };
        process.chdir(flags.target);
        const dataDir = (0, utils_js_1.extensionDataDir)();
        const id = `${manifest.name}`;
        const extensionDir = path.join(dataDir, id);
        const internalSupportDir = (0, utils_js_1.extensionInternalSupportDir)(id);
        const stdoutPath = path.join(internalSupportDir, "stdout.txt");
        const stderrPath = path.join(internalSupportDir, "stderr.txt");
        const pidFile = path.join(internalSupportDir, "cli.pid");
        await Promise.all([extensionDir, internalSupportDir].map((dir) => fsp.mkdir(dir, { recursive: true })));
        await fsp.writeFile(pidFile, `${process.pid}`);
        const outStream = new tail_js_1.Tail(stdoutPath, { forceCreate: true, nLines: 0 });
        const errStream = new tail_js_1.Tail(stderrPath, { forceCreate: true, nLines: 0 });
        outStream.on("line", (data) => {
            logger.logExtensionOut(data.toString());
        });
        outStream.on("error", () => { });
        errStream.on("line", (data) => {
            logger.logExtensionError(data.toString());
        });
        errStream.on("error", () => { });
        await safeBuild(extensionDir);
        process.on("SIGINT", () => {
            logger.logInfo("Shutting down...");
            vicinae.stopDevSession(id);
            throw new Error(`Development session interrupted`);
        });
        const error = vicinae.startDevSession(id);
        if (error) {
            console.error(`Failed to invoke vicinae`, error);
            return;
        }
        chokidar
            .watch(["src", "package.json", "assets"], {
            awaitWriteFinish: { pollInterval: 100, stabilityThreshold: 100 },
            ignoreInitial: true,
        })
            .on("all", async (_, path) => {
            if (path.endsWith("package.json")) {
                manifest = parseManifest();
                logger.logInfo("Generating extension types...");
                (0, extension_types_js_1.updateExtensionTypes)(manifest, flags.target);
            }
            logger.logEvent(`changed file ${path}`);
            await safeBuild(extensionDir);
        });
    }
}
exports.default = Develop;
