// @vicinae/api/dist/api/components/list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.List = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const image_1 = require("../image");
const node_crypto_1 = require("node:crypto");
const metadata_1 = require("./metadata");
const empty_view_1 = require("./empty-view");
const color_1 = require("../color");
const dropdown_1 = require("./dropdown");
function serializeAccessory(accessory) {
    const icon = accessory.icon ? (0, image_1.serializeProtoImage)(accessory.icon) : undefined;
    const tag = typeof accessory === "object" && "tag" in accessory
        ? serializeTag(accessory.tag)
        : undefined;
    const text = typeof accessory === "object" && "text" in accessory
        ? serializeText(accessory.text)
        : undefined;
    return { icon, tooltip: accessory.tooltip, tag, text };
}
function serializeTag(tag) {
    if (tag == null)
        return tag; // null or undefined
    if (typeof tag !== "object")
        return tag;
    if ("color" in tag) {
        const color = (0, color_1.serializeColorLike)(tag.color);
        const value = "value" in tag ? tag.value : undefined;
        return { color, value };
    }
}
function serializeText(text) {
    if (text == null)
        return text; // null or undefined
    if (typeof text !== "object")
        return text;
    if ("color" in text) {
        const color = (0, color_1.serializeColorLike)(text.color);
        const value = "value" in text ? text.value : undefined;
        return { color, value };
    }
}
const ListRoot = ({ searchBarAccessory, children, actions, ...props }) => {
    if (typeof props.enableFiltering === "boolean" &&
        typeof props.filtering === "undefined") {
        props.filtering = props.enableFiltering;
    }
    return ((0, jsx_runtime_1.jsxs)("list", { ...props, children: [searchBarAccessory, children, actions] }));
};
const ListItem = ({ detail, actions, icon, accessories, ...props }) => {
    const id = (0, react_1.useRef)(props.id ?? (0, node_crypto_1.randomUUID)());
    // Icon
    let serializedIcon;
    if (icon && typeof icon === "object" && "value" in icon) {
        serializedIcon = {
            value: icon.value ? (0, image_1.serializeProtoImage)(icon.value) : undefined,
            tooltip: icon.tooltip,
        };
    }
    else {
        serializedIcon = icon ? (0, image_1.serializeProtoImage)(icon) : undefined;
    }
    // Accessories
    const serializedAccessories = accessories?.map(serializeAccessory) ?? undefined;
    return ((0, jsx_runtime_1.jsxs)("list-item", { ...props, icon: serializedIcon, accessories: serializedAccessories, id: id.current, children: [detail, actions] }));
};
const ListItemDetail = ({ metadata, ...props }) => {
    return (0, jsx_runtime_1.jsx)("list-item-detail", { ...props, children: metadata });
};
const ListSection = (props) => {
    return (0, jsx_runtime_1.jsx)("list-section", { ...props });
};
/**
 * @category UI Components
 */
exports.List = Object.assign(ListRoot, {
    Section: ListSection,
    EmptyView: empty_view_1.EmptyView,
    Dropdown: dropdown_1.Dropdown,
    Item: Object.assign(ListItem, {
        Detail: Object.assign(ListItemDetail, {
            Metadata: metadata_1.Metadata,
        }),
    }),
});
