// @vicinae/api/dist/api/components/metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Metadata = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const image_1 = require("../image");
const color_1 = require("../color");
const tag_1 = require("./tag");
const MetadataRoot = (props) => {
    return (0, jsx_runtime_1.jsx)("metadata", { ...props });
};
const MetadataLabel = (props) => {
    const serializedIcon = props.icon
        ? (0, image_1.serializeProtoImage)(props.icon)
        : props.icon;
    let serializedText;
    if (props.text && typeof props.text === "object") {
        serializedText = {
            color: props.text.color ? (0, color_1.serializeColorLike)(props.text.color) : undefined,
            value: props.text.value,
        };
    }
    else {
        serializedText = props.text;
    }
    const { text, ...restProps } = props;
    return (0, jsx_runtime_1.jsx)("metadata-label", { ...restProps, icon: serializedIcon, text: serializedText });
};
const MetadataSeparator = () => {
    return (0, jsx_runtime_1.jsx)("metadata-separator", {});
};
const MetadataLink = ({ title, target, text }) => {
    return (0, jsx_runtime_1.jsx)("metadata-link", { title: title, target: target, text: text });
};
exports.Metadata = Object.assign(MetadataRoot, {
    Label: MetadataLabel,
    Separator: MetadataSeparator,
    TagList: tag_1.TagList,
    Link: MetadataLink,
});
