// @vicinae/api/dist/api/components/form.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Form = exports.DatePickerType = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const use_imperative_form_handle_1 = require("../hooks/use-imperative-form-handle");
const dropdown_1 = require("./dropdown");
const wrapFormItemProps = (props) => {
    return {
        ...props,
        onFocus: () => props.onFocus?.({
            type: "focus",
            target: {
                id: props.id,
                value: (props.value ?? props.defaultValue),
            },
        }),
        onBlur: () => props.onBlur?.({
            type: "blur",
            target: {
                id: props.id,
                value: (props.value ?? props.defaultValue),
            },
        }),
    };
};
const FormRoot = ({ enableDrafts = false, actions, children, isLoading = false, navigationTitle, searchBarAccessory, }) => {
    return ((0, jsx_runtime_1.jsxs)("form", { 
        //@ts-ignore
        enableDrafts: enableDrafts, isLoading: isLoading, navigationTitle: navigationTitle, children: [searchBarAccessory, children, actions] }));
};
const TextField = ({ ref, ...props }) => {
    (0, use_imperative_form_handle_1.useImperativeFormHandle)(ref);
    return (0, jsx_runtime_1.jsx)("text-field", { ...wrapFormItemProps(props) });
};
const PasswordField = ({ ref, ...props }) => {
    (0, use_imperative_form_handle_1.useImperativeFormHandle)(ref);
    return (0, jsx_runtime_1.jsx)("password-field", { ...props });
};
var DatePickerType;
(function (DatePickerType) {
    DatePickerType["DateTime"] = "dateTime";
    DatePickerType["Date"] = "date";
})(DatePickerType || (exports.DatePickerType = DatePickerType = {}));
const DatePickerRoot = ({ ref, onChange, ...props }) => {
    (0, use_imperative_form_handle_1.useImperativeFormHandle)(ref);
    const _onChange = onChange
        ? (newValue) => {
            const dateObj = newValue ? new Date(newValue) : null;
            onChange(dateObj);
        }
        : undefined;
    return ((0, jsx_runtime_1.jsx)("date-picker-field", { ...wrapFormItemProps(props), onChange: _onChange }));
};
const DatePicker = Object.assign(DatePickerRoot, {
    Type: DatePickerType,
    isFullDay: (value) => {
        if (!value)
            return false;
        return (value.getHours() === 0 &&
            value.getMinutes() === 0 &&
            value.getSeconds() === 0 &&
            value.getMilliseconds() === 0);
    },
});
const Checkbox = ({ ref, ...props }) => {
    (0, use_imperative_form_handle_1.useImperativeFormHandle)(ref);
    return (0, jsx_runtime_1.jsx)("checkbox-field", { ...wrapFormItemProps(props) });
};
//FIXME: we probably need to reuse the existing dropdown in
// a smarter way.
const DropdownRoot = ({ ref, children, ...props }) => {
    // FIXME: testing stuff, we need to generalize this to all form items
    (0, use_imperative_form_handle_1.useImperativeFormHandle)(ref);
    return ((0, jsx_runtime_1.jsx)("dropdown-field", { ...wrapFormItemProps(props), children: children }));
};
const Dropdown = Object.assign(DropdownRoot, {
    Item: dropdown_1.Dropdown.Item,
    Section: dropdown_1.Dropdown.Section,
});
const TagPickerRoot = ({ children, ...props }) => {
    return ((0, jsx_runtime_1.jsx)("tag-picker-field", { ...wrapFormItemProps(props), children: children }));
};
const TagPickerItem = ({ icon, ...props }) => {
    return (0, jsx_runtime_1.jsx)("tag-picker-item", { ...props, icon: icon });
};
const TagPicker = Object.assign(TagPickerRoot, {
    Item: TagPickerItem,
});
const TextArea = ({ ref, ...props }) => {
    (0, use_imperative_form_handle_1.useImperativeFormHandle)(ref);
    return (0, jsx_runtime_1.jsx)("text-area-field", { ...wrapFormItemProps(props) });
};
const FilePicker = ({ ref, ...props }) => {
    (0, use_imperative_form_handle_1.useImperativeFormHandle)(ref);
    return (0, jsx_runtime_1.jsx)("file-picker-field", { ...wrapFormItemProps(props) });
};
const Description = (props) => {
    return (0, jsx_runtime_1.jsx)("form-description", { ...props });
};
exports.Form = Object.assign(FormRoot, {
    TextField,
    PasswordField,
    DatePicker,
    Checkbox,
    TextArea,
    Dropdown,
    Description,
    TagPicker,
    FilePicker,
    Separator: () => (0, jsx_runtime_1.jsx)("separator", {}),
    LinkAccessory: ({ target, text }) => ((0, jsx_runtime_1.jsx)("link-accessory", { target: target, text: text })),
});
