// @vicinae/api/dist/api/color.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeColorLike = exports.Color = void 0;
const ui = __importStar(require("./proto/ui"));
/**
 * @category Colors
 */
var Color;
(function (Color) {
    Color["Blue"] = "blue";
    Color["Green"] = "green";
    Color["Magenta"] = "magenta";
    Color["Orange"] = "orange";
    Color["Purple"] = "purple";
    Color["Red"] = "red";
    Color["Yellow"] = "yellow";
    Color["PrimaryText"] = "primary-text";
    Color["SecondaryText"] = "secondary-text";
})(Color || (exports.Color = Color = {}));
const serializeColorLike = (color) => {
    const colorLike = ui.ColorLike.create();
    if (typeof color === "string") {
        colorLike.raw = color;
    }
    else {
        // It's a DynamicColor
        const dynamicColor = ui.DynamicColor.create();
        dynamicColor.light = color.light;
        dynamicColor.dark = color.dark;
        if (color.adjustContrast !== undefined) {
            dynamicColor.adjustContrast = color.adjustContrast;
        }
        colorLike.dynamic = dynamicColor;
    }
    return colorLike;
};
exports.serializeColorLike = serializeColorLike;
