// @vicinae/api/dist/api/local-storage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalStorage = void 0;
const bus_1 = require("./bus");
// Implementation of Raycast's storage API: https://developers.raycast.com/api-reference/storage
/**
 * @category Local Storage
 */
var LocalStorage;
(function (LocalStorage) {
    async function getItem(key) {
        const res = await bus_1.bus.request("storage.get", { key });
        if (!res.ok || res.value.value === null) {
            return undefined;
        }
        return res.value.value;
    }
    LocalStorage.getItem = getItem;
    async function setItem(key, value) {
        await bus_1.bus.request("storage.set", { key, value });
    }
    LocalStorage.setItem = setItem;
    async function removeItem(key) {
        await bus_1.bus.request("storage.remove", { key });
    }
    LocalStorage.removeItem = removeItem;
    async function allItems() {
        const res = await bus_1.bus.request("storage.list", {});
        if (!res.ok)
            return {};
        return res.value.values;
    }
    LocalStorage.allItems = allItems;
    async function clear() {
        await bus_1.bus.request("storage.clear", {});
    }
    LocalStorage.clear = clear;
})(LocalStorage || (exports.LocalStorage = LocalStorage = {}));
